### PR TITLE
Feature/rename

### PR DESCRIPTION
**Pull request by DaQuirm**
This adds `rename/renameSync` as described in:

http://man7.org/linux/man-pages/man2/rename.2.html

### CODE REVIEW

File: packages/test-kit/src/sync-base-fs-contract.ts [hunk 1]
```diff
@@ -409,6 +409,107 @@ export function syncBaseFsContract(testProvider: () => Promise<ITestInput<IBaseF
...
+                const destPath = join(tempDirectoryPath, 'destDir')
+
+                fs.mkdirSync(sourcePath)
+                fs.mkdirSync(destPath)
+                fs.writeFileSync(join(sourcePath, 'file'), SAMPLE_CONTENT)
+
+                expect(() => fs.renameSync(sourcePath, destPath)).not.to.throw('EEXIST')
+            })
+
+            it('updates the parent directory of a renamed entry', () => {
```
Comment (AviVahl) [inline packages/test-kit/src/sync-base-fs-contract.ts]: This test verifies the parent behavior only for a specific fixture (renaming of a directory). Might be useful to `readdirSync` in all other tests (non-error ones) and remove this test. I'm not sure though, WDYT?

File: packages/memory/src/memory-fs.ts [hunk 2]
```diff
@@ -271,6 +272,54 @@ export function createBaseMemoryFsSync(): IBaseMemFileSystemSync {
...
+        }
+
+        delete sourceParentNode.contents[lowerCaseSourceName]
+        sourceNode.name = targetName
+        if (sourceNode.type === 'dir') {
+            Object.getPrototypeOf(sourceNode.contents)['..'] = targetParentNode
+        }
+        targetParentNode.contents[lowerCaseTargetName] = sourceNode
+
+        emitWatchEvent({ path: sourcePath, stats: null })
```
Comment (AviVahl) [inline packages/memory/src/memory-fs.ts]: I wonder which watch events native node's `rename` / `renameSync` triggers. If Node behaves the same across OSs, we could test watch event behavior of rename as well.

File: packages/types/src/base-api.ts [hunk 3]
```diff
@@ -68,6 +68,11 @@ export interface IBaseFileSystemSync {
      * If path is linked, returns the actual target path.
      */
     realpathSync(path: string): string
+
+    /**
+     * Rename (move) a file or a directory
+     */
+    renameSync(filePath: string, newFilePath: string): void
```
Comment (AviVahl) [inline packages/types/src/base-api.ts]: Path might point to a directory so parameters should be `sourcePath` / `destinationPath`.
Also make sure to change all other occurrences of `path` / `newPath` to that as well, for consistency.

**Comment by DaQuirm** [comment]
First of all, that's a very thorough and insightful review, good job! 👍 

> * describe could be separated into "renaming files" and "renaming directories", to be more consistent with other describes. Directory describe would probably contain more tests, due to behaviors with existing target directory (empty or not).

I'd suggest using a separate describe for the directories, since a fair share of tests verifies behavior common for both files and directories (e.g. that the existence of containing directories is checked).

Concerning directory copying: a few relevant cases could not be tested adequately without writing platform-dependent tests. The reason is that there are _still_  discrepancies in the way nodejs `rename` works on different platforms. Consider this open issue for instance: https://github.com/nodejs/node/issues/21957

> * tests currently use `sourcePath`. using `sourceFilePath` / `sourceDirectoryPath` might make following what's being renamed easier. though if describes are separated, this point might be redundant.

That's a good point, it makes things clearer! 👍 

> * after sync tests are reviewed/refactored/ready, please create the async base contract version as well. anyone creating an async-only fs (remote one, for example) should have his `rename` implementation tested too.

Oh, I didn't think about that option! What do you think about making the test code polymorphic in respect of whether it's run synchronously or asynchronously? 

> * should memory's rename touch mtime/birthtime?

It definitely should, totally missed that! 

> * see additional comments next to relevant source lines

Will do!

File: packages/test-kit/src/sync-base-fs-contract.ts [hunk 4]
```diff
@@ -409,6 +409,107 @@ export function syncBaseFsContract(testProvider: () => Promise<ITestInput<IBaseF
...
+                fs.writeFileSync(join(sourcePath, 'file'), SAMPLE_CONTENT)
+
+                fs.renameSync(sourcePath, destinationPath)
+
+                expect(fs.statSync(destinationPath).isDirectory()).to.equal(true)
+                expect(fs.readFileSync(join(destinationPath, 'file'))).to.eql(SAMPLE_CONTENT)
+                expect(() => fs.statSync(sourcePath)).to.throw('ENOENT')
+            })
+
+            it('throws if source path doesn\'t exist', () => {
```
[the remark this answers is not in the document]
Comment (DaQuirm) [reply packages/test-kit/src/sync-base-fs-contract.ts]: Awesome! 😄

File: packages/types/src/base-api.ts (hunk 3 above)
Comment (DaQuirm) [reply packages/types/src/base-api.ts]: That's totally valid, updated the names 👍

File: packages/test-kit/src/sync-base-fs-contract.ts [hunk 5]
```diff
@@ -409,6 +409,107 @@ export function syncBaseFsContract(testProvider: () => Promise<ITestInput<IBaseF
...
+                const { join } = fs.path
+                const sourcePath = join(tempDirectoryPath, 'file')
+
+                fs.writeFileSync(sourcePath, SAMPLE_CONTENT)
+
+                expect(() => fs.renameSync(sourcePath, join(tempDirectoryPath, 'file2'))).not.to.throw('ENOENT')
+                expect(() => fs.renameSync(sourcePath, join(tempDirectoryPath, 'dir', 'file2'))).to.throw('ENOENT')
+            })
+
+            it('doesn\'t throw if destination path already exists for copying a directory over a non-existing directory', () => { //tslint:disable-line
```
[the remark this answers is not in the document]
Comment (DaQuirm) [reply packages/test-kit/src/sync-base-fs-contract.ts]: This testing case defines the desired behavior, other tests may be changed, refactored or even removed, I feel it's good to be explicit about these things but I also understand how it might seem odd. The best way I'd say is to express such things with types whenever possible.

File: packages/test-kit/src/sync-base-fs-contract.ts [hunk 6]
```diff
@@ -409,6 +409,107 @@ export function syncBaseFsContract(testProvider: () => Promise<ITestInput<IBaseF
...
+
+            it('throws if destination containing path doesn\'t exist', () => {
+                const { fs, tempDirectoryPath } = testInput
+                const { join } = fs.path
+                const sourcePath = join(tempDirectoryPath, 'file')
+
+                fs.writeFileSync(sourcePath, SAMPLE_CONTENT)
+
+                expect(() => fs.renameSync(sourcePath, join(tempDirectoryPath, 'file2'))).not.to.throw('ENOENT')
+                expect(() => fs.renameSync(sourcePath, join(tempDirectoryPath, 'dir', 'file2'))).to.throw('ENOENT')
```
[the remark this answers is not in the document]
Comment (DaQuirm) [reply packages/test-kit/src/sync-base-fs-contract.ts]: And it was indeed! Damn, that's a great catch 😺 
Would it make sense to make errors/error handling/error types more elaborate to avoid situations like this?

File: packages/memory/src/memory-fs.ts [hunk 7]
```diff
@@ -271,6 +272,54 @@ export function createBaseMemoryFsSync(): IBaseMemFileSystemSync {
...
+                }
+            } else {
+                throw new Error(`${targetPath} ${FsErrorCodes.PATH_ALREADY_EXISTS}`)
+            }
+        }
+
+        delete sourceParentNode.contents[lowerCaseSourceName]
+        sourceNode.name = targetName
+        if (sourceNode.type === 'dir') {
+            Object.getPrototypeOf(sourceNode.contents)['..'] = targetParentNode
```
[the remark this answers is not in the document]
Comment (DaQuirm) [reply packages/memory/src/memory-fs.ts]: Added a comment, I also agree it makes sense to comment the whole thing or, better yet, replace it with a structural type that support various kinds of entries

File: packages/test-kit/src/sync-base-fs-contract.ts [hunk 8]
```diff
@@ -409,6 +409,107 @@ export function syncBaseFsContract(testProvider: () => Promise<ITestInput<IBaseF
...
+                const { join } = fs.path
+                const sourcePath = join(tempDirectoryPath, 'sourceDir')
+
+                fs.mkdirSync(sourcePath)
+                fs.writeFileSync(join(sourcePath, 'file'), SAMPLE_CONTENT)
+
+                expect(() => fs.renameSync(sourcePath, join(tempDirectoryPath, 'destDir'))).not.to.throw('EEXIST')
+            })
+
+            it('doesn\'t throw if destination path already exists for copying a directory over an empty directory', () => { //tslint:disable-line
```
[the remark this answers is not in the document]
Comment (DaQuirm) [reply packages/test-kit/src/sync-base-fs-contract.ts]: You're right, it sounds better this way. Rephrased all the negativity away 😃

File: packages/test-kit/src/sync-base-fs-contract.ts (hunk 1 above)
Comment (DaQuirm) [reply packages/test-kit/src/sync-base-fs-contract.ts]: I guess this is already indirectly tested for files (by reading them), so this is only to make sure directories stay consistent as well because it can potentially work for files and still get a directory's shadow entries wrong

**Comment by AviVahl** [comment]
Fixes look good. Please check why tests are failing.

I'm not sure how I feel about the new `.sequence()` matcher. It feels a bit duplicate. We could rename `lastEvent()`, and change it to accept either a single event or array of events:

Nice pattern for this:
```ts
public async validateEvents(expectedEvents: IWatchEvent | IWatchEvent[]): Promise<void> {
    const { capturedEvents } = this
    if (!Array.isArray(expectedEvents)) {
        expectedEvents = [ expectedEvents ]
    }
    // typescript knows expectedEvents is an IWatchEvent[] at this point in flow
    ...
}
```
timeout might need to be adjusted (multiplied by `expectedEvents.length`) and so on.

Regarding the non-consistent cross-platform behavior:
- If all platforms throw an error, but different codes, avoid checking it for the specific code. just verify it throws.
- we could add a `boolean` flag to the contract test input to mark specific platforms/fs which we know are diverging. That flag can be used to skip problematic tests.
- If some behavior is too problematic to test cross-platform, skip that test and put a comment explaining why.

Other than that, once you finish cleaning up and everything's green, let me know and I'll squash+release.

**Comment by DaQuirm** [comment]
Okay, watch events proved to be too different to test without resorting to platform-aware code so I'm reverting the test and the changes to `WatchEventsValidator`. This should make the builds green again 💚
